jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/ErrorCloseTest.java

```diff
@@ -0,0 +1,154 @@
+        {
+            container.addMapping("/", (req, resp) -> serverSocket);
+            container.getFactory().addSessionListener(new WebSocketSessionListener() {
+                @Override
+                public void onSessionClosed(WebSocketSession session)
+                {
+                    serverCloseListener.countDown();
+                }
+            });
+        });
+        server.setHandler(contextHandler);
+
+        server.start();
+        client.start();
+    }
+
+    @AfterEach
+    public void stop() throws Exception
+    {
+        client.stop();
+        server.stop();
+    }
+
+    @WebSocket
+    public static class ThrowingSocket extends EventSocket
+    {
+        public List<String> methodsToThrow = new ArrayList<>();
+
+        @Override
+        public void onOpen(Session session)
+        {
+            super.onOpen(session);
+            if (methodsToThrow.contains("onOpen"))
+                throw new RuntimeException("throwing from onOpen");
+        }
+
+        @Override
+        public void onMessage(String message) throws IOException
+        {
+            super.onMessage(message);
+            if (methodsToThrow.contains("onMessage"))
+                throw new RuntimeException("throwing from onMessage");
+        }
+
+        @Override
+        public void onClose(int statusCode, String reason)
+        {
+            super.onClose(statusCode, reason);
+            if (methodsToThrow.contains("onClose"))
+                throw new RuntimeException("throwing from onClose");
+        }
+
+        @Override
+        public void onError(Throwable cause)
+        {
+            super.onError(cause);
+            if (methodsToThrow.contains("onError"))
+                throw new RuntimeException("throwing from onError");
+        }
+    }
+
+    @Test
+    public void testWebSocketThrowsOnTimeout() throws Exception
+    {
+        serverSocket.methodsToThrow.add("onClose");
+        serverSocket.methodsToThrow.add("onError");
+
+        URI uri = new URI("ws://localhost:" + connector.getLocalPort() + "/");
+        EventSocket clientSocket = new EventSocket();
+        Future<Session> connect = client.connect(clientSocket, uri);
+        connect.get(5, TimeUnit.SECONDS);
+
+        // Set a short idleTimeout on the server.
+        assertTrue(serverSocket.open.await(5, TimeUnit.SECONDS));
+        serverSocket.session.setIdleTimeout(1000);
+
+        // Wait for server to timeout.
+        try (StacklessLogging stacklessLogging = new StacklessLogging(WebSocketSession.class))
+        {
+            assertTrue(serverSocket.closed.await(5, TimeUnit.SECONDS));
+        }
+
+        // Check we have stopped the WebSocketSession properly.
+        assertFalse(serverSocket.session.isOpen());
+        serverCloseListener.await(5, TimeUnit.SECONDS);
+        assertTrue(((WebSocketSession)serverSocket.session).isStopped());
+    }
```

Comment:
Am I missing the `onOpen` and `onMessage` version of this test?
Is that done somewhere else?

Reply:
This was really just aimed at the specific case where `onClose` and `onError` throw after a local error like a timeout occurs. Instead of a timeout we could generate the error by throwing from `onOpen` or `onMessage`, but it won't fail like this unless we still throw in `onClose` or `onError`.

Reply:
I will add some more tests for errors from the other methods as well.